runelite-client/src/main/java/net/runelite/client/plugins/worldhopper/WorldHopperPlugin.java

```diff
@@ -0,0 +1,359 @@
+
+	private ScheduledFuture<?> lookupFuture;
+	private WorldResult worldResult;
+
+	private int worldToHopTo = -1;
+
+	@Override
+	protected void startUp() throws Exception
+	{
+		keyManager.registerKeyListener(inputListener);
+
+		lookupFuture = executorService.scheduleAtFixedRate(() ->
+		{
+			try
+			{
+				WorldResult worldResult = new WorldClient().lookupWorlds();
+				worldResult.getWorlds().sort(Comparator.comparingInt(World::getId));
+				this.worldResult = worldResult;
+			}
+			catch (IOException ex)
+			{
+				log.warn(null, ex);
+			}
+		}, 0, 30, TimeUnit.MINUTES);
+	}
+
+	@Override
+	protected void shutDown() throws Exception
+	{
+		keyManager.unregisterKeyListener(inputListener);
+
+		lookupFuture.cancel(true);
+	}
+
+	@Subscribe
+	public void onMenuEntryAdded(MenuEntryAdded event)
+	{
+		int groupId = WidgetInfo.TO_GROUP(event.getActionParam1());
+		String option = event.getOption();
+
+		if (groupId == WidgetInfo.FRIENDS_LIST.getGroupId() || groupId == WidgetInfo.CLAN_CHAT.getGroupId() ||
+			groupId == WidgetInfo.CHATBOX.getGroupId() && !KICK_OPTION.equals(option) || //prevent from adding for Kick option (interferes with the raiding party one)
+			groupId == WidgetInfo.RAIDING_PARTY.getGroupId() || groupId == WidgetInfo.PRIVATE_CHAT_MESSAGE.getGroupId())
+		{
+			boolean after;
+
+			if (AFTER_OPTIONS.contains(option))
+			{
+				after = true;
+			}
+			else if (BEFORE_OPTIONS.contains(option))
+			{
+				after = false;
+			}
+			else
+			{
+				return;
+			}
+
+			// Don't add entry if user is offline
+			ChatPlayer player = getChatPlayerFromName(event.getTarget());
+
+			if (player == null || player.getWorld() == 0 || player.getWorld() == client.getWorld())
+			{
+				return;
+			}
+
+			final MenuEntry hopTo = new MenuEntry();
+			hopTo.setOption(HOP_TO);
+			hopTo.setTarget(event.getTarget());
+			hopTo.setType(MenuAction.RUNELITE.getId());
+			hopTo.setParam0(event.getActionParam0());
+			hopTo.setParam1(event.getActionParam1());
+
+			insertMenuEntry(hopTo, client.getMenuEntries(), after);
+		}
+	}
+
+	private void insertMenuEntry(MenuEntry newEntry, MenuEntry[] entries, boolean after)
+	{
+		MenuEntry[] newMenu = ObjectArrays.concat(entries, newEntry);
+
+		if (after)
+		{
+			int menuEntryCount = newMenu.length;
+			ArrayUtils.swap(newMenu, menuEntryCount - 1, menuEntryCount - 2);
+		}
+
+		client.setMenuEntries(newMenu);
+	}
+
+	@Subscribe
+	public void onHopToMenuClicked(PlayerMenuOptionClicked event)
+	{
+		if (!event.getMenuOption().equals(HOP_TO))
+		{
+			return;
+		}
+
+		ChatPlayer player = getChatPlayerFromName(event.getMenuTarget());
+
+		if (player != null)
+		{
+			hop(player.getWorld());
+		}
+	}
+
+	void hop(boolean previous)
+	{
+		if (client.getGameState() != GameState.LOGGED_IN)
+		{
+			return;
+		}
+
+		int totalLevel = client.getTotalLevel();
+
+		World currentWorld = worldResult.findWorld(client.getWorld());
+
+		// Make it so you always hop out of PVP, high risk and bounty worlds
```

Comment:
What about people aready on PVP worlds? I would assume if someone is already on PVP world, that person wants to hop to another PVP world.

Reply:
I'll check exactly what other clients do tomorrow. I was thinking that quickhopping out of PVP would be something people want, but I'll check out the behaviour of other clients.

For the bounty world, I was thinking that you wouldn't want to hop into a world where a lot of people are PKing if you're in the wilderness, but maybe it's not really that problematic.

Reply:
Some people probably want one and some another, I can definitely see it being useful in both cases, so maybe configuration option.

Reply:
Okay so OSBuddy's quick hopper always hops out of PvP worlds, and never into them. Bounty is regarded as a completely regular world, so BOUNTY will have to be removed from both currentWorld.types and the types of the world that the current is being compared to.

I'll add a config option for enabling only hopping between PvP worlds (if your current world is a pvp world). If you're in a non-PvP world, never hop into one.

Reply:
Changed it to work like OSBuddy's hopper and added a config option for toggling "always hop out of pvp worlds".